database/pebble/db.go

```diff
@@ -0,0 +1,456 @@
+		l.IndexBlockSize = 256 * 1024
+		l.FilterPolicy = bloom.FilterPolicy(10)
+		l.FilterType = pebble.TableFilter
+		if i > 0 {
+			l.TargetFileSize = opts.Levels[i-1].TargetFileSize * 2
+		}
+	}
+	opts.Experimental.ReadSamplingMultiplier = -1 // explicitly disable seek compaction
+
+	log.Info("creating pebbledb")
+
+	db, err := pebble.Open(file, opts)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Database{
+		db:      db,
+		closeCh: make(chan struct{}),
+	}, nil
+}
+
+func (db *Database) Close() error {
+	// close a db twice will trigger panic by pebble instead of error
+	if db.closed.Get() {
+		return database.ErrClosed
+	}
+	db.closed.Set(true)
+
+	err := updateError(db.db.Close())
+	if err != nil && strings.Contains(err.Error(), "leaked iterator") {
+		// avalanche database support close db w/o error
+		// even if there is an iterator which is not released
+		return nil
+	}
+	return err
+}
+
+func (db *Database) HealthCheck(_ context.Context) (interface{}, error) {
+	if db.closed.Get() {
+		return nil, database.ErrClosed
+	}
+	return nil, nil
+}
+
+func (db *Database) Has(key []byte) (bool, error) {
+	if db.closed.Get() {
+		return false, database.ErrClosed
+	}
+
+	_, closer, err := db.db.Get(key)
+	if err == pebble.ErrNotFound {
+		return false, nil
+	}
+
+	if err != nil {
+		return false, updateError(err)
+	}
+	return true, closer.Close()
+}
+
+// Get returns the value the key maps to in the database
+func (db *Database) Get(key []byte) ([]byte, error) {
+	if db.closed.Get() {
+		return nil, database.ErrClosed
+	}
+
+	data, closer, err := db.db.Get(key)
+	if err != nil {
+		return nil, updateError(err)
+	}
+	ret := make([]byte, len(data))
+	copy(ret, data)
+	return ret, closer.Close()
+}
+
+// Put sets the value of the provided key to the provided value
+func (db *Database) Put(key []byte, value []byte) error {
+	// Put causes panic if the db has already been closed
+	if db.closed.Get() {
+		return database.ErrClosed
+	}
+
+	// Use of [pebble.NoSync] here means we don't wait for the [Set] to be
+	// persisted to the WAL before returning. Basic benchmarking indicates that
+	// waiting for the WAL to sync reduces performance by 20%.
+	return updateError(db.db.Set(key, value, pebble.NoSync))
+}
+
+// Delete removes the key from the database
+func (db *Database) Delete(key []byte) error {
+	// Delete causes panic if the db has already been closed
+	if db.closed.Get() {
+		return database.ErrClosed
+	}
+
+	return updateError(db.db.Delete(key, pebble.NoSync))
+}
+
+// If start and limit are the same based on the compare function in pebble
+// pebble returns error but we like to return nil
+func (db *Database) Compact(start []byte, limit []byte) error {
+	// Compact causes panic if the db has already been closed
+	if db.closed.Get() {
+		return database.ErrClosed
+	}
+
+	err := updateError(db.db.Compact(start, limit, true))
+	if err != nil && bytes.Equal(start, limit) {
+		// The default compare function of pebble is bytes.Compare.
```

Comment:
Just to confirm the original compare was  `bytes.Compare(...) == 0` do you have a ref to the original check?